### DISCUSSION TITLE
Toggling "active field" of list of fields

### DISCUSSION THREAD

**Discussion by JMitnik** [react-hook-form/react-hook-form#3811] [Q&A] [answered]
Hi good folks of React Hook Form,

I have a question with a corresponding CSB. It goes like this: 
1. I have a form consisting of a number of `parent properties` and also a list of `children` items.
2. I want to be able to select _one_ of these `children`, and edit them (essentially having an `activeChildIndex` state).
3. This activeChild can be edited in a specific section in my form. I want to jump between children, and edit their details

**My problem**
In the [following Code Sandbox](https://codesandbox.io/s/friendly-meadow-ogzuz?file=/src/App.js), I run into this problem:
1. When I edit one child (for instance Eren), 
2. and then decide I want to edit the other child (thus changing the `activeChild`),
3. => 🔴 the inputs don't reflect this change in activeChild (it still shows `Eren` rather than the other child in the input in the form section).

I am assuming that either my understanding is lacking something (I have been scouring the docs, but I can't entirely be certain if this approach should work or not), or I have to change my approach (for instance, use another useForm and on-submit store the contents of this subform into the children).

Hopefully my question is a bit clear. Thank you either way for looking, and for the amazing work on this library!

(I am aware that this is not performance-optimal, I think I could achieve better performance by putting the `activeChild` form section into its own component and using the `useWatch`).

**Comment by bluebill1049** [comment, marked as answer]
I would try the following option:

1. useForm `shouldUnregister: false`
2. the easier option keeps all the input and show hide their visibility.

let me know if that solves the issue.

**Reply by JMitnik** [reply]
Hi Bill, thanks for the quick response :)!

For clarification, your point 2 refers to hiding the child in css?

Unfortunately, `shouldUnregister:false` does not fix the problem. What happens is that still, when I set the state to a different `activeChildIndex` (so I set active child to not be the one from index 0, but instead index 1), the input with the name `children.${activeChildIndex}.childName` still contains the value from the previous child. As soon as I edit this value, it then changes both children (seems like the input now controls both items of this list). I made a video to show this phenomenon here:

(uploaded in cleanshot, also uploaded in github but it doesn't seem to render so I have an external link)
https://user-images.githubusercontent.com/9025544/103383324-98bd0e80-4af2-11eb-8131-6eda69fd8e2e.mp4


To kind of summarize it: I suppose what I would expect, is that on switching the `activeChildIndex`, the defaultValue of that respective child would be used (rather than using the first rendered child) for the input.

**Reply by JMitnik** [reply]
(also apologies, it seems I put this in RFC accidentally. I put it in Q&A now )

**Reply by JMitnik** [reply]
I think I figured it out. It turns out I should add the `key` property also to the input outside of the `children.map` call I do (this is when working with `useFieldArray`).

I thought that you only need to add `key={field.id}` to the direct child of `fields.map((field, index) => (<div key={field.id} />)` when working with `useFieldArray`, but it turns out also to be important outside of the map (in this case, the inputs in the `activeChild`). 

E.g.

```
const Example = () => {
    // Some index for an active child
    const [activeChildIndex, setActiveChildIndex] = useState();

    const form = useForm({...});
    const { fields } = useFieldArray({
        name: 'children',
        control: form.control,
        keyName: 'id'
    });

    return (
        <div>
            {fields.map((field, index) => (
                // 1. As expected in maps, we use the field's ID as provided here
                <div key={field.id}> 
                    <input name={`children.[${field.id}]`} />
                </div>
            ))}

            {/* ❗❗ 2. BUT ALSO VERY IMPORTANT: If we use a specific chield field fro `useFieldArray` outside of the `.map`, also use the key={id}  */}
            <div className="active-child-field">
                <input name={`children.[${fields[activeChild].id}]`} key={fields[activeChild].id} />
            </div>
        </div>
    )
}
```
